token/program-2022/src/extension/transfer_fee/processor.rs

```diff
@@ -187,18 +181,94 @@ fn process_harvest_withheld_tokens_to_mint(accounts: &[AccountInfo]) -> ProgramR
     let mint_extension = mint.get_extension_mut::<TransferFeeConfig>()?;
 
     for token_account_info in token_account_infos {
-        match harvest_from_account(mint_account_info.key, mint_extension, token_account_info) {
-            // Shouldn't ever happen, but if it does, we don't want to propagate any half-done changes
-            Err(TokenError::Overflow) => return Err(TokenError::Overflow.into()),
+        match harvest_from_account(mint_account_info.key, token_account_info) {
+            Ok(amount) => {
+                let mint_withheld_amount = u64::from(mint_extension.withheld_amount);
+                mint_extension.withheld_amount = mint_withheld_amount
+                    .checked_add(amount)
+                    .ok_or(TokenError::Overflow)?
+                    .into();
+            }
             Err(e) => {
                 msg!("Error harvesting from {}: {}", token_account_info.key, e);
             }
-            Ok(_) => {}
         }
     }
     Ok(())
 }
 
+fn process_withdraw_withheld_tokens_from_accounts(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let mint_account_info = next_account_info(account_info_iter)?;
+    let dest_account_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+    let authority_info_data_len = authority_info.data_len();
+    let account_infos = account_info_iter.as_slice();
+
+    let mint_data = mint_account_info.data.borrow();
+    let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+    let extension = mint.get_extension::<TransferFeeConfig>()?;
+
+    let withdraw_withheld_authority = Option::<Pubkey>::from(extension.withdraw_withheld_authority)
+        .ok_or(TokenError::NoAuthorityExists)?;
+    // This will also iterate over the token accounts for a multisig, but
+    // it will only fail if trying to withdraw from an account that is part
+    // of the multisig, but *not* signing the transaction.
+    // This is an annoyance, but doesn't introduce any security issues.
+    Processor::validate_owner(
+        program_id,
+        &withdraw_withheld_authority,
+        authority_info,
+        authority_info_data_len,
+        account_infos,
+    )?;
+
+    let mut dest_account_data = dest_account_info.data.borrow_mut();
+    let mut dest_account = StateWithExtensionsMut::<Account>::unpack(&mut dest_account_data)?;
+    if dest_account.base.mint != *mint_account_info.key {
+        return Err(TokenError::MintMismatch.into());
+    }
+    if dest_account.base.is_frozen() {
+        return Err(TokenError::AccountFrozen.into());
+    }
+    // This iterates back over any multisig signers, but since it gracefully
+    // fails on those, we can support them.
```

Comment:
Note that this is the goofiest part of the implementation.  I don't want to change the behavior of `validate_owner` for multisigs, so this keeps it simple.  The alternative to iterating through everything is to start at the first non-signer in the `account_infos` slice, but this doesn't resolve the issue mentioned earlier of a multisig whose owner is a token account *not* signing the transaction.

Reply:
Idk, could we avoid some of the iterations in `validator_owner` and goofiness here by passing the number of accounts from which to withdraw as instruction data?

Reply:
That's a great point! For some reason I was thinking that it would be gross to add a new parameter to the instruction creator, but the instruction creator can simply calculate it and tack it on.  And that way, the instruction can create subslices for each part.

Reply:
Ok, done, let me know how it looks

Reply:
That's the idea, but I guess I was thinking of using the field to specify number of withdrawal accounts so that the instruction remains extension-specific and not mixed up in the multisig handling.
It is really the same functionality either way, just feels slightly cleaner to me.

Reply:
Ah right, makes sense. Deep down, I was just avoiding a subtraction which actually makes the code safer. Done!

Reply:
Yeah, I feel you. Subtraction is the worst